sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/PushNotificationClient.java

```diff
@@ -235,133 +225,59 @@ private ChatEvent parsePushNotificationEvent(ChatEventType chatEventType, ChatPu
-                this.registrationErrorHandler.accept(exception);
-            }
-            return;
-        }
-
-        this.logger.info("Scheduling Registrar registration to automatically renew in " + delayMs + " ms");
-
-        TimerTask task = new TimerTask() {
-            @Override
-            public void run() {
-                boolean retry = false;
-
-                PushNotificationClient.this.logger.info("Renew Registrar registration attempt #" + tryCount);
-
-                String skypeUserToken;
-                try {
-                    skypeUserToken = communicationTokenCredential.getToken().get().getToken();
-                    PushNotificationClient.this.refreshEncryptionKeys();
-                    PushNotificationClient.this.registrarClient.register(
-                        skypeUserToken,
-                        deviceRegistrationToken,
-                        PushNotificationClient.this.cryptoKey,
-                        PushNotificationClient.this.authKey);
-                } catch (ExecutionException | InterruptedException e) {
-                    PushNotificationClient.this.logger.logExceptionAsError(
-                        new RuntimeException("Get skype user token for push notification failed: " + e.getMessage()));
-                    retry = true;
-                } catch (Throwable throwable) {
-                    PushNotificationClient.this.logger.logThrowableAsError(throwable);
-                    retry = true;
-                }
-
-                if (retry) {
-                    long delayInMS = 1000L * (long) Math.min(
-                        (int) Math.pow(2.0D, (double) tryCount), NotificationUtils.MAX_REGISTRATION_RETRY_DELAY_S);
-                    PushNotificationClient.this.logger.info("Registration renew failed, will retry in " + delayInMS + " ms");
-                    PushNotificationClient.this.scheduleRegistrationRenew(delayInMS, tryCount + 1);
+    private void startRegistrationRenewalWorker(long intervalInMinutes) {
+        this.logger.info("Initialize RegistrationRenewalWorker in background");
+
+        Data inputData = new Data.Builder().putString("deviceRegistrationToken", deviceRegistrationToken).build();
+
+        PeriodicWorkRequest renewTokenRequest =
+            new PeriodicWorkRequest.Builder(RegistrationRenewalWorker.class, intervalInMinutes, TimeUnit.MINUTES)
+                .setBackoffCriteria(BackoffPolicy.LINEAR, 10, TimeUnit.SECONDS)
+                .setInputData(inputData)
+                .build();
+        workManager.enqueueUniquePeriodicWork("Renewal registration", ExistingPeriodicWorkPolicy.REPLACE, renewTokenRequest);
+
+        //Checking the result of last execution. There are two states for periodic work: ENQUEUED and RUNNING. We do checking
+        //for every ENQUEUED state, meaning last execution has completed.
+        workManager.getWorkInfoByIdLiveData(renewTokenRequest.getId()).observeForever(workInfo -> {
```

Comment:
Does this callback get invoked only if the app is running?

Reply:
How does Contoso app get notified when push notification renewal has failed?

Reply:
It is only invoked when APP is alive. Currently the client is not able to know there is no active registration.

Reply:
What comes to my mind is introducing foreground notification. I think that is the only way to notify clients. But is seems to be over-kill of the problem and I have not analyzed if it is guaranteed to execute in background.

Reply:
Is it possible for Contoso to provide a callback method / class for our Worker class to call directly when renewal has failed?

Reply:
As what has been discussed offline, Contoso will pass the exception hander to the registration worker.

Reply:
Plan to retain this logic to not breaking this API. Will create a new PR adding a new API without the exception handler.